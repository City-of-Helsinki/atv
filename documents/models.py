from django.conf import settings
from django.contrib.postgres.indexes import GinIndex
from django.core.serializers.json import DjangoJSONEncoder
from django.db import models
from django.urls import reverse
from django.utils.translation import gettext_lazy as _
from encrypted_fields import fields

from atv.exceptions import MaximumFileSizeExceededException
from documents.utils import get_attachment_file_path
from documents.validators import BusinessIDValidator
from services.models import Service
from utils.models import TimestampedModel, UUIDModel


class Attachment(TimestampedModel):
    document = models.ForeignKey(
        "Document",
        on_delete=models.CASCADE,
        verbose_name=_("attachment"),
        related_name="attachments",
        help_text=_("Document to which the file is attached."),
    )
    media_type = models.CharField(
        max_length=255,
        default="",
        verbose_name=_("media type"),
        help_text=_("The media type of the attachment."),
    )
    size = models.PositiveIntegerField(
        verbose_name=_("size"),
        help_text=_("Size of the attachment in bytes."),
    )
    filename = models.CharField(
        max_length=255,
        verbose_name=_("filename"),
        help_text=_("The original filename of the attachment."),
    )
    # TODO: The file is not being encrypted yet
    file = models.FileField(
        upload_to=get_attachment_file_path,
        verbose_name=_("file"),
        help_text=_("Encrypted file."),
    )

    class Meta:
        verbose_name = _("attachment")
        verbose_name_plural = _("attachments")
        default_related_name = "attachments"

    def __str__(self):
        return f"Attachment {self.pk}"

    @property
    def uri(self):
        return reverse("documents-attachments-detail", args=[self.document.id, self.id])

    def clean(self):
        if self.file.size > settings.MAX_FILE_SIZE:
            raise MaximumFileSizeExceededException()

    def save(self, *args, **kwargs):
        self.size = self.file.size
        self.filename = self.file.name

        self.full_clean()

        super().save(*args, **kwargs)


class Document(UUIDModel, TimestampedModel):
    service = models.ForeignKey(
        Service,
        on_delete=models.PROTECT,
        verbose_name=_("service"),
        help_text=_("The service which owns this document."),
    )
    user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        blank=True,
        null=True,
        on_delete=models.PROTECT,
        verbose_name=_("user"),
        help_text=_("The creator/owner of this document."),
    )
    business_id = models.CharField(
        max_length=9,
        blank=True,
        validators=[BusinessIDValidator()],
        verbose_name=_("business ID"),
        help_text=_("The business ID of the organization which owns this document."),
    )
    transaction_id = models.CharField(
        max_length=255,
        blank=True,
        verbose_name=_("transaction ID"),
        help_text=_(
            "Transaction identifier given by the owning service. Could be e.g. a UUID."
        ),
    )
    tos_function_id = models.CharField(
        max_length=32,
        verbose_name=_("TOS function ID"),
        help_text=_(
            "UUID without dashes. Should correspond with a Function instance "
            "(e.g. the id from https://api.hel.fi/helerm/v1/function/eb30af1d9d654ebc98287ca25f231bf6/) "
            "which is applied to the stored document when considering storage time."
        ),
    )
    tos_record_id = models.CharField(
        max_length=32,
        verbose_name=_("TOS record ID"),
        help_text=_(
            "UUID without dashes. Should correspond to a record ID "
            "(e.g. records[].id from https://api.hel.fi/helerm/v1/function/eb30af1d9d654ebc98287ca25f231bf6/) "
            "within a Function instance which is applied to the stored document when "
            "considering storage time."
        ),
    )
    draft = models.BooleanField(
        default=False,
        verbose_name=_("draft"),
        help_text=_(
            "Is this document a draft or not. Drafts can be modified by a user."
        ),
    )
    locked_after = models.DateTimeField(
        blank=True,
        null=True,
        verbose_name=_("locked after"),
        help_text=_(
            "Date and time after which this document cannot be modified, except for "
            "deleting. This field should be filled by the calling service if it knows "
            "e.g. that a certain application has a deadline."
        ),
    )
    metadata = models.JSONField(
        default=dict,
        blank=True,
        # null=True,  # TODO should we allow null?
        encoder=DjangoJSONEncoder,
        verbose_name=_("metadata"),
        help_text=_(
            "Key-value pairs given by the calling service. These fields should enable "
            "the service to store some relevant information which it can use to "
            "filter/sort documents, e.g. the handler of the document."
        ),
    )
    content = fields.EncryptedTextField(
        help_text=_("Encrypted content of the document.")
    )
    status = models.CharField(
        max_length=255,
        blank=True,
        help_text=_(
            "Status information given by the owning service. "
            "Could be e.g. some constant string."
        ),
    )
    type = models.CharField(
        max_length=255,
        blank=True,
        verbose_name=_("type"),
        help_text=_(
            "Type information given by the owning service. "
            "Could be e.g. the type of the document."
        ),
    )

    class Meta:
        verbose_name = _("document")
        verbose_name_plural = _("documents")
        default_related_name = "documents"
        indexes = [
            models.Index(fields=["created_at"], name="document_created_at_idx"),
            models.Index(fields=["updated_at"], name="document_updated_at_idx"),
            models.Index(fields=["business_id"], name="document_business_id_idx"),
            models.Index(fields=["transaction_id"], name="document_transaction_id_idx"),
            models.Index(fields=["draft"], name="document_draft_idx"),
            models.Index(fields=["locked_after"], name="document_locked_after_idx"),
            GinIndex(fields=["metadata"], name="document_metadata_idx"),
            models.Index(fields=["status"], name="document_status_idx"),
            models.Index(fields=["type"], name="document_type_idx"),
        ]

    def __str__(self):
        return f"Document {self.pk}"
