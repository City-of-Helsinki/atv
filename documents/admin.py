from django.contrib import admin
from django.urls import reverse
from django.utils.html import format_html
from django.utils.translation import gettext_lazy as _

from documents.models import Attachment, Document


@admin.register(Document)
class DocumentAdmin(admin.ModelAdmin):
    list_display = ("pk", "get_service_name", "draft", "created_at", "updated_at")
    search_fields = ("service__name", "type", "status", "business_id")
    list_filter = ("draft",)
    autocomplete_fields = ("service", "user")
    list_select_related = ("service",)

    @admin.display(description=_("service"), ordering="-service__name")
    def get_service_name(self, obj):
        return obj.service.name


@admin.register(Attachment)
class AttachmentAdmin(admin.ModelAdmin):
    list_display = (
        "pk",
        "filename",
        "media_type",
        "get_size_in_mb",
        "get_document",
        "created_at",
        "updated_at",
    )
    search_fields = (
        "filename",
        "document__id",
    )
    list_filter = ("media_type",)
    autocomplete_fields = ("document",)
    readonly_fields = (
        "filename",
        "media_type",
        "size",
    )

    @admin.display(description=_("size (MB)"))
    def get_size_in_mb(self, obj):
        if not obj.size:
            return "-"
        size_mb = round(float(obj.size) / (1024 ** 2), 2)
        return f"{size_mb}MB"

    @admin.display(description=_("Document"))
    def get_document(self, obj):
        link = reverse("admin:documents_document_change", args=[obj.document.id])
        return format_html('<a href="{}">{}</a>', link, str(obj.document))
