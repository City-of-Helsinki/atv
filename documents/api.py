from rest_framework import status, viewsets
from rest_framework.response import Response


class DocumentViewSet(viewsets.ViewSet):
    def create(self, request):
        return Response({"status": "CREATED"}, status=status.HTTP_201_CREATED)
