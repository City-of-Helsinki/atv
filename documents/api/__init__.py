from .viewsets import AttachmentViewSet, DocumentViewSet

__all__ = [
    "AttachmentViewSet",
    "DocumentViewSet",
]
