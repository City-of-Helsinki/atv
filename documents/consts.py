VALID_OWNER_PATCH_FIELDS = {"draft", "metadata", "content", "attachments"}
