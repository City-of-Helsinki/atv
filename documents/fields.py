import json
from io import BytesIO

from Crypto.Cipher import AES
from django.db import models
from encrypted_fields.fields import EncryptedFieldMixin

from atv.settings import FIELD_ENCRYPTION_KEYS


class EncryptedJSONField(EncryptedFieldMixin, models.JSONField):
    def decrypt(self, value):
        text = super().decrypt(value)
        return json.loads(text)

    def get_db_prep_save(self, value, connection):
        if self.empty_strings_allowed and value == bytes():
            value = ""
        # Instead of using db_prep_value use json.dumps() to get a json formatted object
        value = json.dumps(value)
        if value is not None:
            encrypted_value = self.encrypt(value)
            return connection.Database.Binary(encrypted_value)


class EncryptedFileField(models.FileField):
    def pre_save(self, model_instance, add):
        file = getattr(model_instance, self.attname)
        if file and not file._committed:
            # Commit the file to storage prior to saving the model
            file.save(file.name, self.encrypt_file(file.file), save=False)
        return file

    @staticmethod
    def encrypt_file(file):
        data_to_encrypt = file.read()
        cipher = AES.new(bytes.fromhex(FIELD_ENCRYPTION_KEYS[0]), AES.MODE_GCM)
        nonce = cipher.nonce
        cypher_text, tag = cipher.encrypt_and_digest(data_to_encrypt)
        return BytesIO(nonce + tag + cypher_text)
