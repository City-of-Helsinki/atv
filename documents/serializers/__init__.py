from .attachment import AttachmentSerializer
from .document import CreateAnonymousDocumentSerializer, DocumentSerializer

__all__ = [
    "AttachmentSerializer",
    "CreateAnonymousDocumentSerializer",
    "DocumentSerializer",
]
