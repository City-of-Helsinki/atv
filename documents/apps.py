from django.apps import AppConfig
from django.utils.translation import gettext_lazy as _


class DocumentsConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "documents"
    verbose_name = _("documents")

    def ready(self):
        import documents.signals  # noqa
