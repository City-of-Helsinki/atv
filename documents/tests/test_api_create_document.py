import json
import uuid

import pytest
from django.conf import settings
from django.core.files.uploadedfile import SimpleUploadedFile
from freezegun import freeze_time
from rest_framework import status
from rest_framework.reverse import reverse

from audit_log.models import AuditLogEntry
from documents.models import Attachment, Document

VALID_DOCUMENT_DATA = {
    "status": "handled",
    "type": "mysterious form",
    "transaction_id": "cf0a341b-6bfd-4f59-8d7c-87bf62ba837b",
    "business_id": "1234567-8",
    "tos_function_id": "f917d43aab76420bb2ec53f6684da7f7",
    "tos_record_id": "89837a682b5d410e861f8f3688154163",
    "metadata": json.dumps({"created_by": "alex", "testing": True}),
    "content": json.dumps(
        {
            "formData": {
                "firstName": "Dolph",
                "lastName": "Lundgren",
                "birthDate": "3.11.1957",
            },
            "reasonForApplication": "No reason, just testing",
        }
    ),
}


@freeze_time("2021-06-30T12:00:00+03:00")
def test_create_document(service_api_client, snapshot):
    data = {
        **VALID_DOCUMENT_DATA,
        "attachments": [
            SimpleUploadedFile(
                "document1.pdf", b"file_content", content_type="application/pdf"
            ),
            SimpleUploadedFile(
                "document2.pdf", b"file_content", content_type="application/pdf"
            ),
        ],
    }

    # Patch the document auto-generated UUID to always have the same value
    # to have deterministic tests
    response = service_api_client.post(
        reverse("documents-list"), data, format="multipart"
    )

    assert response.status_code == status.HTTP_201_CREATED
    assert Document.objects.count() == 1
    assert Attachment.objects.count() == 2

    document = Document.objects.first()
    assert document.attachments.count() == 2
    attachment1 = document.attachments.first()
    attachment2 = document.attachments.last()

    body = response.json()
    assert uuid.UUID(body.pop("id")) == document.id
    assert body.pop("attachments", []) == [
        {
            "created_at": "2021-06-30T12:00:00+03:00",
            "filename": "document2.pdf",
            "href": f"http://testserver/v1/documents/{document.id}/attachments/{attachment2.id}/",
            "id": attachment2.id,
            "media_type": "application/pdf",
            "size": 12,
            "updated_at": "2021-06-30T12:00:00+03:00",
        },
        {
            "created_at": "2021-06-30T12:00:00+03:00",
            "filename": "document1.pdf",
            "href": f"http://testserver/v1/documents/{document.id}/attachments/{attachment1.id}/",
            "id": attachment1.id,
            "media_type": "application/pdf",
            "size": 12,
            "updated_at": "2021-06-30T12:00:00+03:00",
        },
    ]
    snapshot.assert_match(body)


@pytest.mark.parametrize(
    "content", ["this shouldn't be allowed", '{"id": 1 "name": "Test"}']
)
def test_create_document_invalid_json_content(service_api_client, content):
    data = {**VALID_DOCUMENT_DATA, "content": content}

    response = service_api_client.post(
        reverse("documents-list"), data, format="multipart"
    )

    assert response.status_code == status.HTTP_400_BAD_REQUEST
    assert Document.objects.count() == 0
    assert Attachment.objects.count() == 0

    body = response.json()
    assert body == {"content": ["Value must be valid JSON."]}


def test_create_document_invalid_fields(service_api_client):
    data = {**VALID_DOCUMENT_DATA, "invalid_field": True}

    response = service_api_client.post(
        reverse("documents-list"), data, format="multipart"
    )

    assert response.status_code == status.HTTP_400_BAD_REQUEST
    assert Document.objects.count() == 0
    assert Attachment.objects.count() == 0

    body = response.json()
    assert body == {"errors": ["Got invalid input fields: invalid_field"]}


def test_create_document_many_files(service_api_client):
    data = {
        **VALID_DOCUMENT_DATA,
        "attachments": [
            SimpleUploadedFile(
                f"document{i}.pdf", b"file_content", content_type="application/pdf"
            )
            for i in range(1, 15)
        ],
    }

    response = service_api_client.post(
        reverse("documents-list"), data, format="multipart"
    )

    assert response.status_code == status.HTTP_400_BAD_REQUEST
    assert Document.objects.count() == 0
    assert Attachment.objects.count() == 0

    body = response.json()
    assert body == {
        "errors": [f"File upload is limited to {settings.MAX_FILE_UPLOAD_ALLOWED}"]
    }


def test_create_document_file_limit(service_api_client):
    data = {
        **VALID_DOCUMENT_DATA,
        "attachments": [
            SimpleUploadedFile(
                "document1.pdf",
                b"x" * (settings.MAX_FILE_SIZE + 1),
                content_type="application/pdf",
            )
        ],
    }

    response = service_api_client.post(
        reverse("documents-list"), data, format="multipart"
    )

    assert response.status_code == status.HTTP_400_BAD_REQUEST
    assert Document.objects.count() == 0
    assert Attachment.objects.count() == 0

    body = response.json()
    assert body == {"errors": ["Cannot upload files larger than 20.0 MB"]}


def test_create_document_no_attachments(service_api_client):
    assert Document.objects.count() == 0
    assert Attachment.objects.count() == 0

    response = service_api_client.post(
        reverse("documents-list"), VALID_DOCUMENT_DATA, format="multipart"
    )

    assert response.status_code == status.HTTP_201_CREATED
    assert Document.objects.count() == 1
    assert Attachment.objects.count() == 0

    body = response.json()
    assert body.get("attachments") == []


@pytest.mark.parametrize("attachments", [0, 1, 2])
def test_audit_log_is_created_when_creating(service_api_client, attachments):
    data = {**VALID_DOCUMENT_DATA}
    if attachments:
        data["attachments"] = [
            SimpleUploadedFile(
                f"document{i}.pdf", b"file_content", content_type="application/pdf"
            )
            for i in range(attachments)
        ]

    response = service_api_client.post(
        reverse("documents-list"), data, format="multipart"
    ).json()

    assert Document.objects.count() == 1

    assert Attachment.objects.count() == attachments
    assert (
        AuditLogEntry.objects.filter(
            message__audit_event__target__type="Document",
            message__audit_event__target__id=response["id"],
            message__audit_event__operation="CREATE",
        ).count()
        == 1
    )
