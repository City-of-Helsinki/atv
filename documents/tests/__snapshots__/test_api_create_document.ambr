# serializer version: 1
# name: test_create_anonymous_document
  dict({
    'business_id': '1234567-8',
    'content': dict({
      'formData': dict({
        'birthDate': '3.11.1957',
        'firstName': 'Dolph',
        'lastName': 'Lundgren',
      }),
      'reasonForApplication': 'No reason, just testing',
    }),
    'content_schema_url': 'https://schema.fi',
    'created_at': '2021-06-30T12:00:00+03:00',
    'deletable': False,
    'delete_after': None,
    'document_language': 'en',
    'draft': False,
    'human_readable_type': dict({
      'en': 'Mysterious Form',
    }),
    'locked_after': None,
    'metadata': dict({
      'created_by': 'alex',
      'testing': True,
    }),
    'service': 'service 34',
    'status': dict({
      'activities': list([
      ]),
      'status_display_values': dict({
        'fi': 'K채sitelty',
      }),
      'timestamp': '2021-06-30T12:00:00+03:00',
      'value': 'handled',
    }),
    'status_histories': list([
      dict({
        'activities': list([
        ]),
        'status_display_values': dict({
          'fi': 'K채sitelty',
        }),
        'timestamp': '2021-06-30T12:00:00+03:00',
        'value': 'handled',
      }),
    ]),
    'tos_function_id': 'f917d43aab76420bb2ec53f6684da7f7',
    'tos_record_id': '89837a682b5d410e861f8f3688154163',
    'transaction_id': 'cf0a341b-6bfd-4f59-8d7c-87bf62ba837b',
    'type': 'mysterious form',
    'updated_at': '2021-06-30T12:00:00+03:00',
    'user_id': None,
  })
# ---
# name: test_create_authenticated_document
  dict({
    'attachments': list([
    ]),
    'business_id': '1234567-8',
    'content': dict({
      'formData': dict({
        'birthDate': '3.11.1957',
        'firstName': 'Dolph',
        'lastName': 'Lundgren',
      }),
      'reasonForApplication': 'No reason, just testing',
    }),
    'content_schema_url': 'https://schema.fi',
    'created_at': '2021-06-30T12:00:00+03:00',
    'deletable': False,
    'delete_after': '2022-12-12',
    'document_language': 'en',
    'draft': False,
    'human_readable_type': dict({
      'en': 'Mysterious Form',
    }),
    'locked_after': None,
    'metadata': dict({
      'created_by': 'alex',
      'testing': True,
    }),
    'service': 'service 36',
    'status': dict({
      'activities': list([
      ]),
      'status_display_values': dict({
        'fi': 'K채sitelty',
      }),
      'timestamp': '2021-06-30T12:00:00+03:00',
      'value': 'handled',
    }),
    'status_histories': list([
      dict({
        'activities': list([
        ]),
        'status_display_values': dict({
          'fi': 'K채sitelty',
        }),
        'timestamp': '2021-06-30T12:00:00+03:00',
        'value': 'handled',
      }),
    ]),
    'tos_function_id': 'f917d43aab76420bb2ec53f6684da7f7',
    'tos_record_id': '89837a682b5d410e861f8f3688154163',
    'transaction_id': 'cf0a341b-6bfd-4f59-8d7c-87bf62ba837b',
    'type': 'mysterious form',
    'updated_at': '2021-06-30T12:00:00+03:00',
  })
# ---
